src/main/java/com/muzika/homeworksb/service/impl/TodoServiceImpl.java

```diff
@@ -0,0 +1,132 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.muzika.homeworksb.dto.TaskHistoryResponseDto;
+import com.muzika.homeworksb.dto.TodoCreateDto;
+import com.muzika.homeworksb.dto.TodoResponseDto;
+import com.muzika.homeworksb.dto.TodoUpdateDto;
+import com.muzika.homeworksb.exception.EntityNotFoundException;
+import com.muzika.homeworksb.mapper.TaskHistoryMapper;
+import com.muzika.homeworksb.mapper.TodoMapper;
+import com.muzika.homeworksb.model.TaskHistory;
+import com.muzika.homeworksb.model.Todo;
+import com.muzika.homeworksb.repository.TaskHistoryRepository;
+import com.muzika.homeworksb.repository.TodoRepository;
+import com.muzika.homeworksb.service.TodoService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.util.List;
+
+@RequiredArgsConstructor
+@Service
+public class TodoServiceImpl implements TodoService {
+
+    private final TodoRepository todoRepository;
+    private final TaskHistoryRepository taskHistoryRepository;
+    private final TodoMapper todoMapper;
+    private final TaskHistoryMapper historyMapper;
+
+    @Autowired
+    private final ObjectMapper objectMapper;
+
+    @Override
+    public TodoResponseDto save(TodoCreateDto createDto) {
+
+        Todo todo = todoMapper.toModel(createDto);
+        // todo.setCreatedDate(LocalDateTime.now()); instead added annotation @CreationTimestamp
+
+        return todoMapper.toDto(
+            todoRepository.save(
+                todo
+            )
+        );
+    }
+
+    @Override
+    public TodoResponseDto findById(Long id) {
+        return todoRepository.findById(id)
+            .map(todoMapper::toDto)
+            .orElseThrow(() -> new EntityNotFoundException("Not found todo by id: " + id));
+    }
+
+    @Override
+    public List<TodoResponseDto> findAll() {
+        return todoRepository.findAll().stream()
+            .map(todoMapper::toDto)
+            .toList();
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        if (todoRepository.existsById(id)) {
+            todoRepository.deleteById(id);
+            return;
+        }
+
+        throw new EntityNotFoundException("Todo with id:" + id + " was not found");
+    }
+
+    @Transactional(rollbackFor = Exception.class, timeout = 1)
+    public TodoResponseDto update(Long id, TodoUpdateDto updateDto) {
+
+        Todo todo = todoRepository.findById(id)
+            .orElseThrow(() -> new EntityNotFoundException("No todo to update by id: " + id));
+
+        String stringifiedOld;
+        String stringifiedNew;
+        try {
+            stringifiedOld = objectMapper.writeValueAsString(todo);
+        } catch (JsonProcessingException e) {
+            System.err.println("Impossible to jsonize todo instance");
```

Comment:
It is better to use `Log4J2` logger instead of `System.out` or `System.err`...

Example:

```java
@SpringBootApplication
public class SpringBootLogbackExtensionsApplication {

    private static final Logger logger = LoggerFactory.getLogger(SpringBootLogbackExtensionsApplication.class);

    public static void main(String[] args) {
        SpringApplication.run(SpringBootLogbackExtensionsApplication.class, args);

        logger.debug("Debug log message");
        logger.info("Info log message");
        logger.error("Error log message");
        logger.warn("Warn log message");
        logger.trace("Trace log message");
    }
}
```


or you can use `@Log4J2` if you are using Lombok